servant/src/Servant/API/Authentication.hs

```diff
@@ -0,0 +1,86 @@
+
+-- | A GADT indexed by policy strictness that encompasses the ways
+-- users will handle the case where authentication data is missing
+-- from a request. For example, suppose we have a Basic-Auth-protected
+-- resource and no appropriate headers are present in the request. Then
+-- the specified 'OnMissing' will be used. In the case it is Lax, then
+-- the servant framework will issue a generalized response. In the case
+-- it is Strict, then api authors can specify how to handle the response.
+-- About the type parameters:
+-- m: the monad errors are retrned in. For now just IO.
+-- e: an error type. For now just ServantErr.
+-- policy: the policy to handle OnMissing.
+data OnMissing m e (policy :: AuthPolicy) where
+    LaxMissing :: OnMissing m e 'Lax
+    StrictMissing :: m e -> OnMissing m e 'Strict
+
+-- | A GADT indexed by policy strictness that encompasses the ways
+-- users will handle the case where the authentication data provided is
+-- rejected. I.e. a username and password do not match in the database.
+-- About the type parameters:
+-- m: the monad errors are returned in. For now just IO.
+-- e: an error type. For now just ServantErr.
+-- authData: the authentication data extracted from the request
+-- errorIndex: an index of error to give a user-provided, semantic meaning to the authentication failure.
+-- policy: the policy to handle OnUnauthenticated actions.
+data OnUnauthenticated m e (policy :: AuthPolicy) errorIndex authData where
+    LaxUnauthenticated :: OnUnauthenticated m e 'Lax errorIndex authData
+    StrictUnauthenticated :: (errorIndex -> authData -> m e) -> OnUnauthenticated m e 'Strict errorIndex authData
+
+-- | A GADT representing the data and functions required to protect a reasource for authentication.
+-- For an authenticated resource, we need to handle the scenario where authentication data is missing
+-- and where authentication data is present but not valid (e.g. uesrname + password not valid).
+-- m: the monad errors are retrned in. For now just IO.
+-- e: an error type. For now just ServantErr.
+-- missingPolicy: the policy to handle missing authentication data actions.
+-- unauthPolicy: the policy to handle rejected authentication attempts.
+-- authData: the type of authData present in a request (e.g. JWT token)
+-- errorIndex: an index of error to give a user-provided, semantic meaning to the authentication failure.
+-- usr: a data type extracted from the authenticated data. This data is likely fetched from a database.
+-- subserver: the rest of the servant API.
+data AuthProtected m e (missingPolicy :: AuthPolicy) (unauthPolicy :: AuthPolicy) errorIndex authData usr subserver =
+    AuthProtected { onMissing :: OnMissing m e missingPolicy
+                  , onUnathenticated :: OnUnauthenticated m e unauthPolicy errorIndex authData
+                  , checkAuth :: authData -> m (Either errorIndex usr)
+                  , subserver :: subserver
+                  }
```

Comment:
@jkarni @alpmestan here is the GADT-based design. There is quite a lot of type-level specification:
- the monad it's run in (`IO` for now)
- the server error (`ServantErr` for now)
- the missing and unauth policy (lets users have strict misisng but lax unauth, for example)
- an `errorIndex`: this was for @jkarni who requested that server writers can provide their own index of errors when trying to authenticate a user (such an index may be: `data ErrorIndex = InvalidUserName | InvalidEmail | InvalidPassword | RateLimited`.
- the `authData` extracted from the request
- the `usr` type fetched from the database
- `subserver`

This proves to be quite flexible. For example, I was able to take the older `Basic` and `JWT`-based auth code and just use a unital index (`()`).

The only annoying part is specifying `IO` and `ServantErr`, but I've exposed combinators in `servant-server` so that clients of the library shouldn't have to worry about that too much.